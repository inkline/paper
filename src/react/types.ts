export type VNode = JSX.Element;

export interface ComponentContext {
    useSlot: (name: string) => void;
}

export type ConstructorType =
    | NumberConstructor
    | StringConstructor
    | BooleanConstructor
    | ObjectConstructor
    | ArrayConstructor;

export type ComponentProps<Props> = {
    [key in keyof Props]: ConstructorType | {
        type: ConstructorType;
        default?: Props[key] | (() => Props[key]);
    };
}

export interface ComponentDefinition<Props, State> {
    props?: ComponentProps<Props>;
    setup?(props: Props): State;
    render(state: Props & State, context: ComponentContext): VNode;
}
