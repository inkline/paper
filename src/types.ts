export type Ref<T> = { value: T };

export type UnwrapState<State> = {
    [key in keyof State]: (State[key] & Ref<State[key]>)['value']
};

export type ConstructorType =
    | NumberConstructor
    | StringConstructor
    | BooleanConstructor
    | ObjectConstructor
    | ArrayConstructor;

export type ComponentProps<Props> = {
    [key in keyof Props]: ConstructorType | {
        type?: ConstructorType;
        default?: Props[key] | (() => Props[key]);
    };
}
