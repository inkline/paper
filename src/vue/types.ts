import { VNode } from 'vue';
import { ComponentProps } from '@inkline/ucd/types';

export interface ComponentContext {
    slots: {
        [key: string]: () => Element
    };
    attrs: Record<string, any>;
    emit: (...args: any[]) => void
}

export interface ComponentDefinition<Props, State> {
    props?: ComponentProps<Props>;
    setup?(props: Props, context: ComponentContext): State;
    render(state: Props & State, context: ComponentContext): VNode;
}
