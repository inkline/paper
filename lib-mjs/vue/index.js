export * from '@inkline/paper/vue/computed';
export * from '@inkline/paper/vue/defineComponent';
export * from '@inkline/paper/vue/events';
export * from '@inkline/paper/vue/h';
export * from '@inkline/paper/vue/provide';
export * from '@inkline/paper/vue/ref';
export * from '@inkline/paper/vue/types';
//# sourceMappingURL=index.js.map