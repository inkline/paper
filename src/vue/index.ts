export { computed, h, ref } from 'vue';
export * from '@inkline/ucd/vue/defineComponent';
export * from '@inkline/ucd/vue/events';
export * from '@inkline/ucd/vue/types';
