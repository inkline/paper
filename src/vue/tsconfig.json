{
    "extends": "../../tsconfig.json",
    "compilerOptions": {
        "types": [
            "vue"
        ]
    }
}
