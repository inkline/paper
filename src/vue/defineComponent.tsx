import { SetupContext } from 'vue';
import { ComponentDefinition } from '@inkline/ucd/vue';

export function defineComponent<Props, State> (definition: ComponentDefinition<Props, State>) {
    return {
        props: definition.props || {},
        setup (props: Props, { slots, emit }: SetupContext) {
            const ctx = {
                useSlot (name: string) {
                    return ctx.slots[name]!();
                },
                emit
            };

            const state = definition.setup
                ? { ...props, ...definition.setup(props, ctx) }
                : props as Props & State;

            return () => definition.render(state, ctx);
        }
    };
}
