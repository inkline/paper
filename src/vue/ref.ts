import { ref } from 'vue';
export {
    ref
};
