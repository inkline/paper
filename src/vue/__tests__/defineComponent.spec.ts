import { render, fireEvent } from '@testing-library/vue';
import { defineComponent, h, ref } from '@inkline/ucd/vue';
import { Ref } from '@inkline/ucd/types';

describe('vue', () => {
    describe('defineComponent()', () => {
        it('should return the component definition itself', () => {
            const componentDefinition = defineComponent({
                render () {
                    return h('div');
                }
            });

            expect(componentDefinition).toEqual(componentDefinition);
        });

        it('should render a vue component with state', () => {
            const Component = defineComponent<{}, { count: Ref<number>; }>({
                setup () {
                    const count = ref(0);

                    return { count };
                },
                render (state) {
                    return h('div', {}, [
                        `${state.count.value}`
                    ]);
                }
            });

            const wrapper = render(Component);
            expect(wrapper.html()).toMatchSnapshot();
        });

        it('should render a vue component with state and props', () => {
            const Component = defineComponent<{ initialValue: number; }, { count: Ref<number>; }>({
                props: {
                    initialValue: {
                        type: Number,
                        default: 0
                    }
                },
                setup (props) {
                    const count = ref(props.initialValue);

                    return { count };
                },
                render (state) {
                    return h('div', {}, [
                        `${state.count.value}`
                    ]);
                }
            });

            const wrapper = render(Component, {
                props: {
                    initialValue: 1
                }
            });
            expect(wrapper.html()).toMatchSnapshot();
        });

        it('should render a vue component with state, props, and onClick', async () => {
            const Component = defineComponent<{ initialValue: number; }, { count: Ref<number>; onClick(): void }>({
                props: {
                    initialValue: {
                        type: Number,
                        default: 0
                    }
                },
                setup (props) {
                    const count = ref(props.initialValue);

                    const onClick = () => {
                        count.value += 1;
                    };

                    return { count, onClick };
                },
                render (state) {
                    return h('button', { onClick: () => state.onClick() }, [
                        `${state.count.value}`
                    ]);
                }
            });

            const wrapper = render(Component, {
                props: {
                    initialValue: 1
                }
            });
            await fireEvent.click(wrapper.container.firstChild as Element);
            expect(wrapper.html()).toMatchSnapshot();
        });
    });
});
