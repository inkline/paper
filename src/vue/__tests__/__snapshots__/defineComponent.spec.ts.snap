// Vitest Snapshot v1

exports[`vue > defineComponent() > should render a vue component with state 1`] = `"<div>0</div>"`;

exports[`vue > defineComponent() > should render a vue component with state and props 1`] = `"<div>1</div>"`;

exports[`vue > defineComponent() > should render a vue component with state, props, and onClick 1`] = `"<button>2</button>"`;
